src/vehicles/RosVehicle.lua

```diff
@@ -0,0 +1,257 @@
+    if self:getLaserFrameNode() then
+        for i = 0, spec.laser_scan_obj.vehicle_table.laser_scan.num_layers-1 do
+            spec.laser_scan_array = {}
+            -- get the (world) coordinate of each laser scanner's origin: (orig_x, orig_y, orig_z)
+            -- "laser_dy" is added between the scanning planes, along +y direction (locally) from the lowest laser scan plane
+            -- and all laser scan planes are parallel to each other
+            local laser_dy = spec.laser_scan_obj.vehicle_table.laser_scan.inter_layer_distance * i
+            local orig_x, orig_y, orig_z = localToWorld(spec.LaserFrameNode, 0, laser_dy, 0)
+
+            for j = 0, (spec.laser_scan_obj.vehicle_table.laser_scan.num_rays - 1) do
+                local seg_theta = j * spec.laser_scan_obj.delta_theta
+                -- (i_laser_dx, 0 , i_laser_dz) is a local space direction to define the world space raycasting (scanning) direction
+                local i_laser_dx = -spec.laser_scan_obj.sin(seg_theta) * spec.laser_scan_obj.radius
+                local i_laser_dz = -spec.laser_scan_obj.cos(seg_theta) * spec.laser_scan_obj.radius
+                local dx, dy, dz = localDirectionToWorld(spec.LaserFrameNode, i_laser_dx, 0 , i_laser_dz)
+                spec.laser_scan_obj:getLaserData(spec.laser_scan_array, orig_x, orig_y, orig_z, dx, dy, dz)
+            end
+
+            -- FS time is "frozen" within a single call to update(..), so this
+            -- will assign the same stamp to all LaserScan messages
+            local t = ros_time
+
+            -- create LaserScan instance
+            local scan_msg = sensor_msgs_LaserScan.new()
+
+            -- populate fields (not using LaserScan:set(..) here as this is much
+            -- more readable than a long list of anonymous args)
+            scan_msg.header.frame_id = spec.base_link_frame .."/laser_frame_" .. i
+            scan_msg.header.stamp = t
+            -- with zero angle being forward along the x axis, scanning fov +-180 degrees
+            scan_msg.angle_min = spec.laser_scan_obj.vehicle_table.laser_scan.angle_min
+            scan_msg.angle_max = spec.laser_scan_obj.vehicle_table.laser_scan.angle_max
+            scan_msg.angle_increment = spec.laser_scan_obj.LS_FOV / spec.laser_scan_obj.vehicle_table.laser_scan.num_rays
+            -- assume sensor gives 50 scans per second
+            scan_msg.time_increment = (1.0 / 50) / spec.laser_scan_obj.vehicle_table.laser_scan.num_rays
+            --scan_msg.scan_time = 0.0  -- we don't set this field (TODO: should we?)
+            scan_msg.range_min = spec.laser_scan_obj.vehicle_table.laser_scan.range_min
+            scan_msg.range_max = spec.laser_scan_obj.vehicle_table.laser_scan.range_max
+            scan_msg.ranges = spec.laser_scan_array
+            --scan_msg.intensities = {}  -- we don't set this field (TODO: should we?)
+
+            -- publish the message
+            pub_scan:publish(scan_msg)
```

Comment:
Could this be moved to the new `LaserScan` class?

Perhaps in a `do_scan(..)` method or something?

Reply:
Yes, I've moved it to the `LaserScanner` class